### ISSUE TITLE
Pyright using incorrect type definition for variable (Structural typing inappropriately applying)

### ISSUE THREAD

**Issue by BrandonNav** [microsoft/pyright#5180]
**Describe the bug**
When type checking a variable of a type that nearly matches another type in a pyi file, Pyright is using the type from the pyi file rather than the type in the file.

**To Reproduce**
1. In a new virtualenv: `pip install django-stubs==4.2`
2. Create `test.py` with contents below
3. Run pyright

```bash
npx pyright@latest test.py
```

**Background context**

django-stubs has the following file: `django-stubs/contrib/sites/requests.pyi`

```python
class HttpRequest(BytesIO):
    GET: _ImmutableQueryDict
    POST: _ImmutableQueryDict
    COOKIES: dict[str, str]
    META: dict[str, Any]
    ... # etc

class _MutableHttpRequest(HttpRequest):
    GET: QueryDict  # type: ignore[assignment]
    POST: QueryDict  # type: ignore[assignment]
```

**Code**
I have the following code in test.py

```python
# pyright: basic
from __future__ import annotations
from django.http import HttpRequest

class SitemapRequest(HttpRequest):
    GET: dict
    META: dict
    
def get_sitemap_request() -> SitemapRequest:
    request = SitemapRequest() # reveal_type(request) == _MutableHttpRequest
    return request
```

**Expected behavior**
I would expect request to have a type of `SiteMapRequest`. `_MutableHttpRequest` and `SiteMapRequest` are almost structurally the same and inherit from `HttpRequest` but nothing is a Protocol so I would expect nominal typing to occur rather than structural

**VS Code extension or command-line**
Python: 3.10.11
Pyright: 1.1.310
Pylance: v2023.5.41

Occurs both on CLI and Pylance

**Comment by erictraut** [comment]
This is because of a bug in the `django-stubs` stub library. The `HttpRequest` class (defined in `django-stubs/http/request.pyi`) declares a `__new__` method that has a return type of `_MutableHttpRequest`. It should be `Self` if this class is intended to be used with subclasses. If you manually change it to `Self` in the stub, you will get the desired behavior. 

You can also work around the issue by defining a `__new__` method in your own code for the derived class `SitemapRequest`.

```python
class SitemapRequest(HttpRequest):
    if TYPE_CHECKING:
        def __new__(cls) -> Self:
            ...
```

I recommend reporting this bug to the maintainers of `django-stubs`.

I recently reworked the logic in pyright for evaluating constructor calls. There were a bunch of bugs in the previous implementation where it wasn't properly honoring return types for metaclass `__call__` methods or `__new__` methods. In other words, it wasn't accurately reflecting the way that constructors work at runtime. It's possible that the `django-stubs` bug was previously masked by these bugs in pyright. That could explain why this hasn't been reported previously.

Incidentally, mypy shares some of the same behaviors that I recently fixed in pyright. There are open bugs in the mypy issue tracker on this topic.

**Comment by BrandonNav** [comment]
Thanks for the info and the work around @erictraut 
I'll close this and ask django-stubs to update the stubs
